### PR TITLE
Control what gets installed as node_modules/.bin/ scripts with allow-scripts

### PR DESCRIPTION
**Pull request by naugtur**
Prevent bin script confusion (https://socket.dev/blog/npm-bin-script-confusion) with @lavamoat/allow-scripts.

What is the new feature here?
- add configuration to turn off linking up bin scripts (warning: no way to do that in yarn3 yet, will have to come up with a workaround in the yarn plugin)
- generate policies for selecting which bin scripts are installed
- enforce those policies when linking up bin scripts the default npm way with their own implementation (should match yarn1 behavior too)
- add a script that errors out and prints instructions on how to enable a script in place of all other scripts in case they're being called

no yarn3 support yet, but it might end up being a big old `rm` added to the yarn3 plugin

Note:
This implementation cares about (and guards) nested bins as well. They might be used by lifecycle scripts of dependencies.

Initial setup of a pre-configured repo is going to have to depend on a package.json->scripts entry with a direct path to an allow-scripts cli. That way allow-scripts remains operational even if not in the allow list.

----

- Bin confusion protections are now experimental and not enabled by default, so we can test them better in real projects
- All the mess in this branch's history is meant to be squashed

### CODE REVIEW

**Comment by kumavis** [comment]
`yarn test:prep` should auto generate the policies for allow-scripts

**Comment by naugtur** [comment]
> `yarn test:prep` should auto generate the policies for allow-scripts

@kumavis there is no test:prep in allow-scripts

File: packages/allow-scripts/test/index.js
Comment (naugtur) [inline packages/allow-scripts/test/index.js]: This test now also documents how bin scripts work - the nesting and the availability in PATH

File: packages/allow-scripts/test/projects/.gitignore
Comment (legobeat) [inline packages/allow-scripts/test/projects/.gitignore]:
```suggestion
.goodscriptworked

```

Comment (legobeat) [reply packages/allow-scripts/test/projects/.gitignore]: ...not sure about that suggestion. Either way; missing newline at EOF.

Comment (naugtur) [reply packages/allow-scripts/test/projects/.gitignore]: fixed

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js]: Any reason not to process these concurrently?

```suggestion
  return Promise.all(allowedBins.map(
    ({ bin, path, link, canonicalName }) => {
      console.log(`- ${canonicalName}`)
      return linkBinRelative({ path, bin, link, force: true })
    }
  )
```

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js:221]: See above

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js]:
```suggestion
  console.log('\n# allowed packages with lifecycle scripts')
```

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js]:
```suggestion
    console.log('\n# unconfigured packages with lifecycle scripts')
```

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js]:
```suggestion
  console.log('\n# disallowed packages with lifecycle scripts')
```

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js]:
```suggestion
    console.log('\n# packages with lifecycle scripts that no longer need configuration due to package or scripts removal')
```

File: packages/allow-scripts/src/whichbin.js
Comment (legobeat) [inline packages/allow-scripts/src/whichbin.js]: Assuming the env var from https://github.com/LavaMoat/LavaMoat/blob/1fbcc2e49979ae5269f9053129da368fb3e90dbe/packages/allow-scripts/src/setup.js#L16 is intended to be used for project selection
```suggestion
  const rootDir = process.env.INIT_CWD || process.cwd()
```

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js]:
```suggestion
      const binsList = Object.entries(normalizeBin(depPackageJson)?.bin || {})
```

File: packages/allow-scripts/src/index.js
Comment (naugtur) [reply packages/allow-scripts/src/index.js]: In case of an error, the names in output would not help figure out which one failed. 
And they do run in parallel within linkBinRelative AFAIR when there's more than one bin in a package.

File: packages/allow-scripts/src/index.js
Comment (naugtur) [reply packages/allow-scripts/src/index.js]: I forgot it's supported in node14

File: packages/allow-scripts/README.md
Comment (legobeat) [inline packages/allow-scripts/README.md]:
```suggestion
- `auto` will generate an allowlist of top-level bin scripts allowed for execution
```

File: packages/allow-scripts/README.md
Comment (legobeat) [inline packages/allow-scripts/README.md]:
```suggestion
- `setup` will add a new configuration option to your project package manager RC file to disable linking up bin scripts
```

File: packages/allow-scripts/README.md
Comment (legobeat) [inline packages/allow-scripts/README.md]:
```suggestion
- `run` will link up the allowed scripts and replace not allowed scripts with an error
```

File: packages/allow-scripts/README.md
Comment (legobeat) [inline packages/allow-scripts/README.md]:
```suggestion
When you attempt to run a bin script not in the allowlist, you will get an error with instructions on how to enable it manually.
```

File: packages/allow-scripts/src/toggles.js
Comment (legobeat) [inline packages/allow-scripts/src/toggles.js]: Can this get a more descriptive name?
It would also be great if we could if not initialize default value then at least document the interface `{bins: boolean}`

(Also: newline at EOF)

File: packages/allow-scripts/src/setup.js
Comment (legobeat) [inline packages/allow-scripts/src/setup.js:10]: I note some overlap and differences with new implementation for #387 so I guess this might see some structural changes soon again?

(Still seems good to not overthink it, to consider this in isolation, and take that as it becomes necessary)

File: packages/allow-scripts/src/setup.js
Comment (legobeat) [inline packages/allow-scripts/src/setup.js:125]: I note some overlap and differences with new implementation for https://github.com/LavaMoat/LavaMoat/issues/387 so I guess this might see some structural changes soon again?

(Still seems good to not overthink it, to consider this in isolation, and take that in the following changesets as it becomes necessary, which I guess is the intention here?)

https://github.com/LavaMoat/LavaMoat/compare/main...lava-cli#diff-b64c06b15a9e440c63fed29404849b37a115750c524c7a632e160681a0392290R77

File: packages/allow-scripts/src/setup.js
Comment (legobeat) [inline packages/allow-scripts/src/setup.js]:
```suggestion
```

File: packages/allow-scripts/test/index.js
Comment (legobeat) [inline packages/allow-scripts/test/index.js:84]:
```suggestion
    force: true,
```

File: packages/allow-scripts/test/index.js
Comment (legobeat) [inline packages/allow-scripts/test/index.js:119]:
```suggestion
    force: true,
```

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js:145]:
```suggestion
    console.log('\nconfiguration looks good as is - no changes necessary')
```

File: packages/allow-scripts/src/index.js
Comment (legobeat) [reply packages/allow-scripts/src/index.js]: Hm, I don't see how they run in parallel there - something additional in `bin-links`?

But either way I see the point on the message ordering - sounds reasonable to make that take priority :+1:

File: packages/allow-scripts/README.md
Comment (legobeat) [inline packages/allow-scripts/README.md]:
```suggestion
Bin script confusion is a new attack where a dependency gets its script to run by declaring executables that end up on the path and later get triggered either by the user or by other programs. More details in [npm bin script confusion: Abusing ‘bin’ to hijack ‘node’ command](https://socket.dev/blog/npm-bin-script-confusion) by Socket.dev
```

File: packages/allow-scripts/README.md
Comment (legobeat) [inline packages/allow-scripts/README.md]:
```suggestion
To enable protection against bin script confusion, run all of the above `allow-scripts` commands with the `--experimental-bins` flag.
```

File: packages/allow-scripts/src/cli.js
Comment (legobeat) [inline packages/allow-scripts/src/cli.js]:
```suggestion
      describe: 'opt-in to set up experimental protection against bin script confusion',
```

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js:381]: Fix or remove..?

File: packages/allow-scripts/src/index.js
Comment (legobeat) [inline packages/allow-scripts/src/index.js:95]: I think this really shouldn't be taken lightly - modifying package manager rc files and package.json is one thing but deleting files (that are typically in .gitignore) is something that should not be done without a capitalized "Are you sure?" or explicit `--force-delete-bins` flag IMO... Users may have setups and expectations we have no idea about that would make this a bad surprise.

If anything: Recreate in a new temp location, diff with current, error/warn about diffs

**Comment by legobeat** [comment]
Aside from what's raised in comments (that is, most of the actual code): LGTM

(also: needs rebase on `main`)

File: packages/allow-scripts/README.md
Comment (naugtur) [reply packages/allow-scripts/README.md]: that's the first time it slipped in my writing in a few months. surprising :D

File: packages/allow-scripts/src/index.js
Comment (naugtur) [reply packages/allow-scripts/src/index.js:95]: This note originates from the fact that yarn3 doesn't have a setting to disable bin linking and we'll have to do it for yarn3 somehow if we want full control over bin links at all times.
The other pkgmanagers get configured to leave the .bin empty and we're running with that assumption. Existing items in node_modules/.bin will get overwritten anyway (I explored an idea that'd put stuff there before running install, and all pkgmanagers destroyed what I created unless blocked on filesystem level requiring root access)

I'm considering this step as a backup in case there was a bypass in existing protections. if the assumption is .bin doesn't exist, deleting it preemtively seems like a good idea...

File: packages/allow-scripts/src/index.js
Comment (naugtur) [reply packages/allow-scripts/src/index.js:381]: This was broken for a long time. I didn't want to fix it in this feature.
Left it as an eyesore so that I'm more likely to investigate what feature this was supporting many refactorings ago. Looking at this is meant to hurt :smiling_imp:

File: packages/allow-scripts/src/index.js
Comment (legobeat) [reply packages/allow-scripts/src/index.js:95]: Hm, if the case really is that all (at least supported) package managers wipe it regardless, I guess it should be all right, but maybe in a future change regardless?

File: packages/allow-scripts/src/setup.js
Comment (naugtur) [reply packages/allow-scripts/src/setup.js:125]: This change happened first, but I stopped myself from further structural changes to limit the conflict later ;) 

Once this lands, I'll rebase lava-cli and arrange for this added functionality to use the tooling over there.

File: packages/allow-scripts/src/toggles.js
Comment (naugtur) [reply packages/allow-scripts/src/toggles.js]: I stopped myself too early, yes.

My favorite format of feature toggle module is a function that accepts:
```js
featureToggle('feature-name', {
on: ()=>{},
off: ()=>{}
})
```
And demands all keys in use to be explicitly set before use.
I thought it'd be a bit of an overkill. Gotta decide where on the specrtum to stop ;)   
Counting your comment as "not _this_ early".

Comment (legobeat) [reply packages/allow-scripts/src/toggles.js]: My 5c: Something like this keeps the current simplicity and makes intentions more clear. Some jsdoc typings would make it sufficient? 

```
module.exports = {
  FeatureToggled: {
    bins: undefined, // or if you want to put default value here already
  },
};
```

File: packages/allow-scripts/src/index.js
Comment (naugtur) [reply packages/allow-scripts/src/index.js:95]: yes, in a future change, as we refine the logic against real usage.

**Comment by naugtur** [comment]
Rebase done, will make the feature toggle prettier soon, with no meaningful change to logic to avoid undoing the LGTM :smile:

File: packages/allow-scripts/src/setup.js
Comment (naugtur) [inline packages/allow-scripts/src/setup.js]: windows?

Comment (naugtur) [reply packages/allow-scripts/src/setup.js]: linking it up in here could be a good alternative

**Review by kumavis — APPROVED** [review]
thanks for the walkthrough on the endo call/

File: packages/allow-scripts/src/setup.js
Comment (naugtur) [reply packages/allow-scripts/src/setup.js]: `node ./unixpath` works on windows

File: packages/allow-scripts/src/toggles.js
Comment (naugtur) [reply packages/allow-scripts/src/toggles.js]: I put the default value in and made sure all toggles need to be defined upfront in the file. how do you like that?

Comment (legobeat) [reply packages/allow-scripts/src/toggles.js]: Much better, and nice usage of `Object.seal` :)

**Review by legobeat — APPROVED** [review]

**Pull request merged**
